metabase/metabase
# add hide button to column formatting popover in table

Closes https://github.com/metabase/metabase/issues/28608

### Description

We are adding a `Hide this column` button to the column settings popover in the `Table` visualization.

Enabling this functionality is a bit tricky. Normally, these settings widgets only have access to the `qb.card.visualization_settings.column_settings` field in our Redux store. However, the `enabled` field that controls visibility of a column is actually in a difference slice, `qb.card.visualization_settings["table.columns"]`.

One potential workaround would have been to define another `enabled` field under `column_settings` for each column. However, this would probably end up becoming inconvenient, as there would be two sources of truth for the `enabled` state, thus anywhere we need to know if a column is `enabled` would require us to write two checks there, checking the columns in `column_settings` and in `table.columns`.

Instead, I decided to read and update `"table.columns"` directly from the settings widget for the `Hide this column` button. This required grabbing that field from the Redux store using `useSelector` and writing a bespoke `useHideColumn` hook to provide a function to perform the update.

### How to verify

1. Open a table visualization (either open a model or table directly, or create a question)
2. Click on a column header -> gear icon -> "Hide this column"
3. Press the play button to refresh to table visualization, and the column should be gone

### Demo

https://user-images.githubusercontent.com/37751258/224402378-2a9be203-7050-46d1-9b41-7bcfd5378778.mov

### Checklist

- ✅ Tests have been added/updated to cover changes in this PR
